### ISSUE TITLE
Round Trip DateTime not working

### ISSUE THREAD

**Issue by NetTecture** [object/Simple.OData.Client#508]
I have a model I define as (I count out the irrelevant stuff).

    public class Account
    {

        [Key]
        public string Identity { get; set; }

        public DateTime Created { get; set; }

    }

As er spec read properties should be ignored and / or round tripped. So, I set up a new class like that:

        [TestMethod]
        public async Task Post()
        {
            var c = new ODataClient(TestSetup.ClientSettings);

// Removed
            var account = new OdataApi.Account()
            {
            };
            var insertedAccount = await c.For<OdataApi.Account>()
                .Set(account)
                .InsertEntryAsync();
        }

Note that DateTime Created is not initialized. This may lead to a problem, but should not be relevant to this question.

For round tripping, this DateTime must be sent as DateTimeOffset within a defined Timezone (how do I set that? Should be in OdataClientSettings).

The post gets the following error:

Result StackTrace:	
at Simple.OData.Client.V4.Adapter.RequestWriter.GetPropertyValue(IEdmTypeReference propertyType, Object value, ODataResource root)
   at Simple.OData.Client.V4.Adapter.RequestWriter.<>c__DisplayClass21_0.<CreateODataEntry>b__6(KeyValuePair`2 x)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Simple.OData.Client.V4.Adapter.RequestWriter.CreateODataEntry(String typeName, IDictionary`2 properties, ODataResource root)
   at Simple.OData.Client.V4.Adapter.RequestWriter.WriteEntryContentAsync(String method, String collection, String commandText, IDictionary`2 entryData, Boolean resultRequired)
   at Simple.OData.Client.RequestWriterBase.CreateInsertRequestAsync(String collection, String commandText, IDictionary`2 entryData, Boolean resultRequired)
   at Simple.OData.Client.RequestBuilder.InsertRequestAsync(Boolean resultRequired, CancellationToken cancellationToken)
   at Simple.OData.Client.ODataClient.InsertEntryAsync(FluentCommand command, Boolean resultRequired, CancellationToken cancellationToken)
   at Simple.OData.Client.BoundClient`1.InsertEntryAsync(Boolean resultRequired, CancellationToken cancellationToken)
   at Tests.Api.Odata.Web.Controllers.AccountTests.Post() in C:\Work\OBB\Source.Backend\Tests.Api.Odata.Web\Controllers\AccountTests.cs:line 61

This is not good. What is the reommended solution here? I generally hve no problem using Edm.DateTimeOffset but would really like not to pollute the API model with external references that pull in significant additional references. Now, Attributes are out (somehow the OData teams thinks it is funny to pull in the whole aspnetcore framework in - nice), but it looks like Microsoft.OData.Edm is self-contained.

**Comment by object** [comment]
What is the exact error POST triggers? You pasted stacktrace, but what is the actual exception?

**Comment by NetTecture** [comment]
Ups.

Result Message:	
Test method Tests.Api.Odata.Web.Controllers.AccountTests.Post threw exception: 
System.NotSupportedException: Conversion is not supported from type System.DateTime to OData type [Edm.DateTimeOffset Nullable=False Precision=0]

**Comment by object** [comment]
So your service metadata specifies DateTimeOffset, and you are sending DateTime. In general you need to provider additional information in order for conversion to be correct, see here for example:
https://docs.microsoft.com/en-us/dotnet/standard/datetime/converting-between-datetime-and-offset

What can probably be done is to specify some hint in ODataClientSettings about how such conversion should be performed.

**Comment by NetTecture** [comment]
Well, the problem is that given that there is no DateTime, the conversion in the metadata is automatic - this is the way asp.net webapi odata has decided to support DateTime on the C# side (no joke, it s***s). Anyway, in Odata I have to decide on the timezone, which I do:

            app.UseMvc(r => {
                r.SetTimeZoneInfo(TimeZoneInfo.Utc);

I have the classes in a separate project ;)

So, reading works fine, but writing does not and I really would love to use SImple.OData.Web for testing which involves round tripping.

Basically, what is needed (technically) is:

* You need to determine what happens with a DateTime on the way out, which may involve defining a DefaultTimeZoneINnfo ;)

THen you can do the conversion.

Regards

THomas

**Comment by object** [comment]
I just pushed version 5.5.0-beta1 that is supposed to fix this issue.

**Comment by NetTecture** [comment]
Ah, sorry, that got overlooked. Not working. Do I have to configure it?

            var c = new ODataClient(TestSetup.ClientSettings);
            var identity = await c
                .For<OdataApi.User>()
                .Top(1)
                .FindEntryAsync();
            var account = new OdataApi.Account()
            {
                Identity = Guid.NewGuid(),
                AccountType = OdataApi.AccountTypes.ServiceProvider,
                User = identity,
                Created = DateTime.UtcNow,
            };
            var insertedAccount = await c.For<OdataApi.Account>()
                .Set(account)
                .InsertEntryAsync();

still returns Message: Test method Tests.Api.Odata.Web.Controllers.AccountTests.Post threw exception: 
System.NotSupportedException: Conversion is not supported from type System.DateTime to OData type [Edm.DateTimeOffset Nullable=False Precision=0]

**Comment by object** [comment]
@NetTecture this is strange, you don't need to configure anything. Have a look at these new tests:

https://github.com/object/Simple.OData.Client/blob/v5/src/Simple.OData.Client.UnitTests/Core/RequestWriterTests.cs

You will find tests that set BirthDate and HireDate (the latter is defined as Edm.Date). And when setting BirthDate I use both DateTime with Unspecified kind and DateTimeOffset. Can you see if anything there is different from your case?

**Comment by danielabbatt** [comment]
This doesn't work when the DateTime value is DateTime.MinValue and the client is in a timezone ahead of UTC
